### PR TITLE
implement zeroed and uninitialized with MaybeUninit

### LINKED ISSUE #62825: re-land #62150 (Implement mem::{zeroed,uninitialized} in terms of MaybeUninit)
#62150 got reverted to give users some time to migrate; we should re-land it.

**Original issue:**

https://github.com/rust-lang/rust/pull/62150 has changed the behavior of `mem::uninitialized()` (and `zeroed()`), causing a few popular crates to be killed by `SIGILL`. A writeup can be found [here](https://github.com/amethyst/amethyst/issues/1808#issuecomment-513449537).

There are cases where outdated (EOL) versions of downstream dependencies have this issue. In that case, the upstream crate will have to migrate to a newer major version of the dependency, which means it will have to deal with those breaking changes.

Therefore, I suggest we back out the change for at least one stable release, providing the community with the time to upgrade.

cc @RalfJung

### PR DESCRIPTION
**Pull request by RalfJung**
This is the second attempt of doing such a change (first PR: https://github.com/rust-lang/rust/pull/62150). The last change [got reverted](https://github.com/rust-lang/rust/pull/63343) because it [caused](https://github.com/rust-lang/rust/issues/62825) some [issues](https://github.com/rust-lang/rust/issues/52898#issuecomment-512182438) in [code that incorrectly used these functions](https://github.com/erlepereira/x11-rs/issues/99).

Since then, the [problematic code has been fixed](https://github.com/erlepereira/x11-rs/pull/101), and rustc [gained a lint](https://github.com/rust-lang/rust/pull/63346) that is able to detect many misuses of these functions statically and a [dynamic check that panics](https://github.com/rust-lang/rust/pull/66059) instead of causing UB for some incorrect uses.

Fixes https://github.com/rust-lang/rust/issues/62825

### CODE REVIEW

File: src/test/ui/intrinsics/intrinsic-move-val.rs [hunk 1]
```diff
@@ -15,17 +14,17 @@ pub fn main() {
...
 
         // A normal overwrite, to demonstrate `check_drops_state`.
-        x = box D(2);
+        x = Some(box D(2));
 
         // At this point, one destructor has run, because the
         // overwrite of `x` drops its initial value.
         check_drops_state(1, Some(1));
 
-        let mut y: Box<D> = rusti::init();
```
Comment (oli-obk) [inline src/test/ui/intrinsics/intrinsic-move-val.rs]: Oh no... did we have UB in our test suite?

Comment (RalfJung) [reply src/test/ui/intrinsics/intrinsic-move-val.rs]: Yes we did indeed. And this circumvented my lint by using the intrinsic directly.

Well, that's not possible any more after this PR. :)

**Comment by RalfJung** [comment]
@bors p=0
I'd prefer my other PR #69999 to go first.

**Comment by RalfJung** [comment]
Giving back priority (https://github.com/rust-lang/rust/pull/69922#issuecomment-598499342)
@bors p=1

File: src/test/ui/init-large-type.rs [hunk 2]
```diff
@@ -10,17 +10,13 @@
...
-}
+use std::{mem, thread};
 
 const SIZE: usize = 1024 * 1024;
 
 fn main() {
     // do the test in a new thread to avoid (spurious?) stack overflows
     thread::spawn(|| {
-        let _memory: [u8; SIZE] = unsafe { init() };
+        let _memory: [u8; SIZE] = unsafe { mem::zeroed() };
```
Comment (RalfJung) [inline src/test/ui/init-large-type.rs]: This test is failing on the "noopt" builder. And that makes sense; there's now a function call "in the way" here where previously the intrinsic was expanded in-line to zero-initializing LLVM IR.

**Comment by RalfJung** [comment]
@oli-obk I added `-O` to the init-large-type test as that test now clearly needs inlining to work. Does this make sense?

**Comment by oli-obk** [comment]
Maybe we should mark the wrapper functions as `#[inline(always)]`? Is there any reason for them to never get inlined?

**Comment by RalfJung** [comment]
Sure, I can do that. Note that this already applied before this PR.

**Comment by oli-obk** [comment]
Oh, right. so it's the more complex code in MaybeUninit that llvm doesn't handle without optimizations. I guess that's alright, though maybe we can teach mir optimizations about that at some point.

@bors r+

**Comment by Centril** [comment]
@bors retry rollup=maybe

Ignoring `rollup=never` directive due to constrained CI budget & combining with orthogonal PRs in rollup.

**Comment by RalfJung** [comment]
"Received request to deprovision: The request was cancelled by the remote provider."...?

@bors retry

**Pull request merged**